examples/vello_editor/src/text.rs

```diff
@@ -126,7 +140,28 @@ impl Editor {
         // TODO: support clipboard on Android
     }
 
-    pub fn handle_event(&mut self, event: &WindowEvent) {
+    pub fn handle_event(&mut self, window: &Window, event: &WindowEvent) {
+        if let ComposeState::Preedit { text_at } = self.compose_state {
+            // Clear old preedit state when handling events that potentially mutate text/selection.
+            // This is a bit overzealous, e.g., pressing and releasing shift probably shouldnt't
+            // clear the preedit.
+            if matches!(
+                event,
+                WindowEvent::KeyboardInput { .. }
+                    | WindowEvent::MouseInput { .. }
+                    | WindowEvent::Ime(..)
```

Comment:
If you remove `::Ime` from here, I think that might be what Aaron meant?

Reply:
The previous preedit state has to be cleared (or otherwise handled) while handling an `Ime::Preedit`. In `Ime::Commit` it's technically not necessary, because Winit guarantees it clears the preedit just prior to a commit.

I started working on a somewhat more performant version where the preedit text is not simply wiped on IME events. Rather, on subsequent preedit events I find the common prefix and update the `Editor::buffer` field accordingly, usually resulting in one fewer memory shift of the editor text (e.g., preedit changing from "n" to "ni", where we don't first have to remove the "n", which shifts everything following to the left), and sometimes no shifts (e.g., preedit changing from "a" to "b", which can be done in-place).

State management becomes more complicated though, so I postponed it until the details have been worked out. Fcitx5 seems to send a clearing preedit between preedits anyway, so there it wouldn't make a practical difference. There (probably?) are IMEs that behave differently.

Reply:
I think the main point is that the only case where you need to remove the previous preedit is when a new pre-edit arrives.

Reply:
Different applications handle this differently.

In Firefox when you're entering with IME and then do something else (for example, removing focus from the text field, or changing cursor location within the field by pressing with the pointer), the preedit text is inserted as-is. Firefox also seems to get into a weird state sometimes: after this sequence, starting a new IME input often fails to display preedit text.

In LibreOffice Writer, when you change cursor location while a preedit is shown, the preedit is moved to the new cursor anchor. If you move focus away from the main text area, the preedit is still shown, but when you move focus back, the moment you do anything, the preedit is removed.

Winit correctly keeps the IME state alive when other events are sent; so when changing cursor location, I think we can move the preedit text to the new location, and it will be updated correctly when the user continues typing. It appears we can reset the IME's state when a text field loses focus by sending `Window::set_ime_allowed(false)` (Winit sends an `IME::Disabled` in response.)

Reply:
Hmm, that all matches my understanding, but wasn't really the point I was trying to make. Given the state of this example, i.e. containing a single text field, winit expects you to only change the preedit in response to `Ime::Preedit`.

When changing the cursor position, it really isn't clear what the best behaviour is, although Firefox definitely gets it wrong on my machine.

Reply:
Ah, I see. I was talking about the behavior because stopping the preedit from being displayed was a result of not knowing what the correct behavior should be.

I've since changed the behavior so the preedit text is only touched in response to `Ime::Preedit` and `Ime::Disabled`.